HelperFunctions/uitools/uigetcode.m

```diff
@@ -0,0 +1,159 @@
+%       function name is returned.
+%   
+%   * Be sure to include newlines and the "%" on each line if you want it
+%       appear as a MATLAB comment.
+% Outputs:
+%   code: char array of code entered by user
+%   fn_name: the function name set by the user if allowed, or the same as
+%       the input name given. This is only validated as a valid name if the
+%       user chooses (e.g. edit_fn_name = true input).
+
+% Note, this has a pretty high overhead for a relatively simple task; but I
+%   couldn't find documented or undocumented techniques to get the simple
+%   editor that is used when you want to edit a datatip for example.
+
+if nargin < 3
+    beginning_text = '';
+end
+if nargin < 4
+    ending_text = '';
+end
+if nargin < 5
+    code = '';
+end
+if nargin < 6
+    edit_fn_name = false;
+end
+char_inputs = {'path','fn_name','beginning_text','ending_text','code'};
+for i = 1:length(char_inputs)
+    if ~eval(sprintf('ischar(%s)',char_inputs{i}))
+        error('"%s" should be a char vector.',char_inputs{i});
+    end
+end
+assert(islogical(edit_fn_name),'"edit_fn_name" should be a scalar logical.');
+
+% Define contents from bottom up
+post = [newline 'end' newline ending_text newline newline,...
+        '%#ok<*FNDEF> Ignores filename/function name mismatch.'];
+fn = [fn_name '()' newline];
+pre =  [beginning_text newline,...
+        '% When finished, you can save and close this window to continue',newline,...
+        '%    (don''t change the filename or path)', newline];
+if edit_fn_name
+    pre = [pre ,...
+        '%    (add a function name)', newline,...
+        '%    (edit inside the function)', newline newline,...
+        'function '];
+    % Keep fn separate from pre
+    init = [pre, fn, code, newline, post];
+else
+    pre = [pre ,...
+        '%    (only edit inside the function)', newline newline,...
+        'function ' fn]; % Add fn to pre
+    init = [pre, code, newline, post];
+end
+nPre = length(pre);
+nPost = length(post);
+
+doc = matlab.desktop.editor.findOpenDocument(path);
+if isempty(doc)
+    doc = matlab.desktop.editor.newDocument(init);
+    if exist(path,'dir')==7
+        mkdir(path);
+    end
+    doc.saveAs(path);
+else % Reset
+    doc.Text = init;
+end
+% Prepare cursor
+lastGoodSpot = nPre;
+doc.JavaEditor.setCaretPosition(lastGoodSpot);
+if edit_fn_name % Highlight placeholder
+    [line, start] = doc.indexToPositionInLine(lastGoodSpot);
+    doc.Selection = [line start line start+length(fn_name)];
+end
+doc.makeActive();
+
+while doc.Opened
+    % Jail cursor to within function (note use could slip outside
+    % between lines of this while loop; these are separate threads.
+    try
+        % NOTE: len and cursor are NOT atomic, so we will check if it
+        % changed while cursor updated.
+        len = doc.JavaEditor.getLength();
+        cursor = doc.JavaEditor.getCaretPosition();
+        if doc.JavaEditor.getLength() ~= len; continue; end % Typing/cutting/pasting
```

Comment:
While this looks silly, it is 100% necessary for multi-threaded applications that don't have thread locks.